Issue #43 Spec doc for CDI injection.

manovotn opened the pull request:
Two things to note here are that default configured instances are not shared (which IMO follows the result of our previous discussions) unless you name them and that I deliberately left out any mention of custom producers as the possibility of that is simply implied by using CDI as a mechanism.

@njr-11 here is my take on the CDI part. I had it mostly completed before but wanted to do some polishing before sending it. Please let me know what you think about it.

genie-microprofile commented:
Can one of the admins verify this patch?

manovotn commented:
I'll look into the ECA compliance (and other things around that) first thing tomorrow.

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+
+=== Injecting `ThreadContext`
+
+Application can inject instances of `ThreadContext` using CDI with the option to share same `ThreadContext` between multiple injection points.
```
Similar to comment on Injecting Manager, this sentence is confusing where currently placed (introducing an example of not sharing).  I'd suggest moving it to the section on "Sharing `ThreadContext` Instances", but you already covered this information there, so it seems best to just omit from the current section.

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+=== Injecting `ThreadContext`
+
+Application can inject instances of `ThreadContext` using CDI with the option to share same `ThreadContext` between multiple injection points.
+The simplest use case will create a new `ThreadContext` instance for every such injection point and the configuration will be equal to default values of `@ThreadContextConfig` annotation.
```
Minor grammar fixes, similar to the ManagedExecutor equivalent sentence,

The simplest use case is a new `ThreadContext` instance per injection point, where the configuration is equivalent to the default values of the @ThreadContextConfig annotation.

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+
+==== Configuring `ThreadContext`
+
+Oftentimes there is need to fine-grain the `ThreadContext` instance and modify what contexts are to be propagated.
```
Oftentimes there is a need to configure the `ThreadContext` instance to specify which contexts to propagate.

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+==== Configuring `ThreadContext`
+
+Oftentimes there is need to fine-grain the `ThreadContext` instance and modify what contexts are to be propagated.
+This is where `@ThreadContextConfig` annotation comes into play allowing to configure all the aspects of an injected `ThreadContext` instance.
```
The `@ThreadContextConfig` annotation configures all aspects of an injected `ThreadContext` instance.

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+
+Oftentimes there is need to fine-grain the `ThreadContext` instance and modify what contexts are to be propagated.
+This is where `@ThreadContextConfig` annotation comes into play allowing to configure all the aspects of an injected `ThreadContext` instance.
+Note that using just this annotation, new instance of `ThreadContext` is created for every injection point even should those injection points have equal configuration.
```
When using this annotation without specifying the `@NamedInstance` qualifier, a new instance of `ThreadContext` is injected for every such injection point, even if those injection points have equivalent configuration.

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+==== Sharing `ThreadContext` Instances
+
+Last but not least, sometimes it is desirable to share the same instance of `ThreadContext` among several injection points.
+For such use case, the specification provides a CDI qualifier `@NamedInstance` which takes one parameter - a unique name assigned to the given `ThreadContext` configuration.
```
Recommend rewriting the first two sentences as:

The CDI qualifier, `@NamedInstance`, enables you to share the same instance of `ThreadContext` across multiple injection points. The `@NamedInstance` qualifier has a single parameter which specifies the unique name that is assigned to the given `ThreadContext` configuration.

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+
+Last but not least, sometimes it is desirable to share the same instance of `ThreadContext` among several injection points.
+For such use case, the specification provides a CDI qualifier `@NamedInstance` which takes one parameter - a unique name assigned to the given `ThreadContext` configuration.
+Having one injection point that defines both, the name using `@NamedInstance` and the configuration using `@ThreadContextConfig`, application can then inject the same instance repeatedly using the qualifier.
```
When an injection point is annotated with both `@ThreadContextConfig` and `@NamedInstance`, the container creates a `ThreadContext` instance matching the configuration and injects this same instance into all injection points within the application that are annotated with a matching `@NamedInstance` qualifier.

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+}
+----
+
+Since the configuration of `ThreadContext` is optional, a shared `ThreadContext` with default configuration is also a possibility.
```
Similar to the earlier pattern, this seems wrong. In the example that follows, the `@NamedInstance("sharedDefaultTC")` qualifier does not match any instance, so wouldn't this be an injection failure?

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
 ----
 
-The container provides default instances of `ManagedExecutor` and `ThreadContext`, as injected above, which are shared within an application with all other injection points for `ManagedExecutor` and `ThreadContext` that do not specify a qualifier. Their configuration is equivalent to `ManagedExecutor.builder().build()` and `ThreadContext.builder().build()`.
+The container allows to inject default instances of `ManagedExecutor` and `ThreadContext` which do not leverave either `@NamedInstance` or `ManagedExecutorConfig`/`ThreadContextConfig` annotations, as shown above.
```
In the absence of other qualifiers and annotations, the container creates and injects a new default instance of `ManagedExecutor` or `ThreadContext` per injection point, as shown above.

aguibert replied:
I am in favor of having `@Default` injection points use a shared instance, rather than one instance per injection point.  Conceptually I think of `@NamedInstance` as qualifier, and `@XXConfig` being a "synthetic producer" for the associated qualifier (whether the qualifier is explicitly declared or inferred via FQCN+field).

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
 
-The container provides default instances of `ManagedExecutor` and `ThreadContext`, as injected above, which are shared within an application with all other injection points for `ManagedExecutor` and `ThreadContext` that do not specify a qualifier. Their configuration is equivalent to `ManagedExecutor.builder().build()` and `ThreadContext.builder().build()`.
+The container allows to inject default instances of `ManagedExecutor` and `ThreadContext` which do not leverave either `@NamedInstance` or `ManagedExecutorConfig`/`ThreadContextConfig` annotations, as shown above.
+These instances are not shared and container creates a new instance for each injection point of `ManagedExecutor` or `ThreadContext`.
```
Usage of the term 'shared' here is slightly confusing.  It is shared with the same injection point (for example, if in a request scoped bean that is created/destroyed often), but not with other injection points.  That said, given the requested update to the previous sentence, the current sentence should now be redundant and can be omitted.

manovotn replied:
Agreed, that's a good reword you proposed

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
-The container provides default instances of `ManagedExecutor` and `ThreadContext`, as injected above, which are shared within an application with all other injection points for `ManagedExecutor` and `ThreadContext` that do not specify a qualifier. Their configuration is equivalent to `ManagedExecutor.builder().build()` and `ThreadContext.builder().build()`.
+The container allows to inject default instances of `ManagedExecutor` and `ThreadContext` which do not leverave either `@NamedInstance` or `ManagedExecutorConfig`/`ThreadContextConfig` annotations, as shown above.
+These instances are not shared and container creates a new instance for each injection point of `ManagedExecutor` or `ThreadContext`.
+The configuration of these instances is equal to default values present in `@ManagedExecutorConfig` and `ThreadContextConfig`.
```
The configuration of these instances is equivalent to the default values of `@ManagedExecutorConfig` and `@ThreadContextConfig`.

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
 === Injection of Configured Instances
 
-TODO write this section after discussion about the Config annotations is resolved.
+The `@ManagedExecutorConfig` and `@ThreadContextConfig` provide the ability to configure the injection points with desired capabilities.
```
The `@ManagedExecutorConfig` and `@ThreadContextConfig` annotations provide the ability to configure the injection points with desired capabilities.

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
+
+=== Sharing Configured Instances
+
+Configured instances of `ManagedExecutor` and `ThreadContext` can also be shared so long as they are given a CDI qualifier `@NamedInstance` which assigns them a unique name per configuration.
```
Configured instances of `ManagedExecutor` and `ThreadContext` are shared by their unique name, which is assigned by the CDI qualifier, `@NamedInstance`.

manovotn replied:
`are shared by their unique` sounds a bit weird.
I'd go for `are shared based on their unique name, which is...`

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
+=== Sharing Configured Instances
+
+Configured instances of `ManagedExecutor` and `ThreadContext` can also be shared so long as they are given a CDI qualifier `@NamedInstance` which assigns them a unique name per configuration.
+Given that one injection point contains a configuration and a name, application can inject the same instance elsewhere using just the CDI qualifier as an identifier.
```
When one injection point is annotated with both a configuration and a `@NamedInstance` qualifier, the application can inject the same instance elsewhere by annotating other injection points with just the matching qualifier.

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
+
+Configured instances of `ManagedExecutor` and `ThreadContext` can also be shared so long as they are given a CDI qualifier `@NamedInstance` which assigns them a unique name per configuration.
+Given that one injection point contains a configuration and a name, application can inject the same instance elsewhere using just the CDI qualifier as an identifier.
+In the example below, `executor1` injection point defines a configuration and a name `myExec` for a configured `ManagerExecutor` instance.
```
In the example below, the `executor1` injection point defines a configuration and a name, `myExec`, for a configured `ManagerExecutor` instance.

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
+Configured instances of `ManagedExecutor` and `ThreadContext` can also be shared so long as they are given a CDI qualifier `@NamedInstance` which assigns them a unique name per configuration.
+Given that one injection point contains a configuration and a name, application can inject the same instance elsewhere using just the CDI qualifier as an identifier.
+In the example below, `executor1` injection point defines a configuration and a name `myExec` for a configured `ManagerExecutor` instance.
+The other injection points, `executor2` and `executor3`, then get the same `ManagedExecutor` instance by using the `@NamedInstance` qualifier with the name `myExec`.
```
The other injection points, `executor2` and `executor3`, share the same `ManagedExecutor` instance by using the `@NamedInstance` qualifier with the name `myExec`.

njr-11 commented on spec/src/main/asciidoc/overview.asciidoc:
```diff
@@ -79,12 +79,52 @@ Instances of `ManagedExecutor` and `ThreadContext` can be injected into CDI bean
+    ... // in some other bean
+    @Inject
+    void setCompletableFuture(@NamedInstance("myExec") ManagedExecutor executor2) {
+        completableFuture = exec.newIncompleteFuture();
```
completableFuture = executor2.newIncompleteFuture();

aguibert commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+=== Injecting `ManagedExecutor`
+
+Application can inject instances of `ManagedExecutor` using CDI with the option to share same executors between multiple injection points.
+The simplest use case will create a new executor instance for every such injection point and the configuration will be equal to default values of `@ManagedExecutorConfig` annotation.
```
(This ties into one of the later comments I left about treating the default `@ManagedExecutorConfig` annotation differently than the absence of the annotation entirely, it would be helpful to read that comment first)

If we stick with the pattern that just `@Inject @NamedInstance` declares  _usage_, and `@Inject @NamedInstance @ManagedExecutorConfig` declares _usage and definition_, I think it would make more sense to have 1 shared `ManagedExecutor` across all default injection points (`@Inject ManagedExecutor exec;`).

Since the default ManagedExecutor instance is unbounded (infinite maxAsnyc and infinite maxQueued), it doesn't really matter if we share these instances or not, but conceptually it makes more sense to only get a dedicated instances when you are using `@ManagedExecutorConfig` (which is analogous to a Producer).  

What do others think on this?

manovotn replied:
> Since the default ManagedExecutor instance is unbounded 

That's what we designed now, but note that we discussed https://github.com/eclipse/microprofile-concurrency/issues/31 and external config of default settings. With that in place, defaults could differ (depending on what we allow there, but I see no reason not to permit changing all parameters).

>  I think it would make more sense to have 1 shared ManagedExecutor across all default injection points

Like I said (elsewhere on this PR), this approach is the opposite to what we drafted for sharing. The argument back there was that sharing by default is not intuitive for user. Hence sharing defaults is non-intuitive as well, especially given that in all other cases it behaves differently.

Impl-wise this isn't a problem. You can detect injection points that have a name but no config linked to them. In such case you simply provide default with a name attached to it to enable sharing.

aguibert commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+public class FooBean {
+
+  @Inject
+  @NamedInstance("sharedDefaultTC")
```
same comment here as with ManagedExecutors.  Also, if we are going to show examples of invalid code in our spec doc, we should make it more obvious that the code example is bad, or point out via in-line comments what is wrong with the code example.  
Reading this, I'm not certain if the following text about a DefinitionException is describing this code example or not.

manovotn replied:
> Reading this, I'm not certain if the following text about a DefinitionException is describing this code example or not.

All proposed text describes how to achieve a certain behaviour and then shows it with bits of code and this is no exception. Besides this code won't throw the exception because similar configuration is not given a different name.

njr-11 replied:
Note: This falls under the same category as the similar section for ManagedExecutor that will be discussed under the main issue

manovotn commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
\ No newline at end of file
+=== Injecting `ManagedExecutor`
+
+Application can inject instances of `ManagedExecutor` using CDI with the option to share same executors between multiple injection points.
```
I originally wanted to give an overview of options, but we might as well drop it.

manovotn commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+
+Oftentimes there is need to fine-grain the executor instance and change the number of maximum asynchronous tasks or define what contexts are to be propagated.
+This is where `@ManagedExecutorConfig` annotation comes into play allowing to configure all the aspects of an injected `ManagedExecutor` instance.
+Note that using just this annotation, new instance of `ManagedExecutor` is created for every injection point even should those injection points have equal configuration.
```
I didn't want to get ahead of myself and mention @NamedInstance before actually explaining it. I would prefer to keep it that way. I'll reword the ending as suggested though.

manovotn commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+
+Last but not least, sometimes it is desirable to share the same instance of `ManagedExecutor` among several injection points.
+For such use case, the specification provides a CDI qualifier `@NamedInstance` which takes one parameter - a unique name assigned to the given executor configuration.
+Having one injection point that defines both, the name using `@NamedInstance` and the configuration using `@ManagedExecutorConfig`, application can then inject the same instance repeatedly using the qualifier.
```
I wanted to define this more from a user perspective, e.g. that you need to have one IP with name and config and the rest with just config. That the container provides shared bean is an "impl detail".
Nonetheless, will alter the wording to improve it.

manovotn commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,158 @@
+}
+----
+
+Since the configuration of `ManagedExecutor` is optional, a shared `ManagedExecutor` with default configuration is also a possibility.
```
Not necessarily, we said that you can `@Inject ManagedExecutor` without config and it has default settings.
Furthermore we said (actually you fought for this to be default) that **sharing is not default behaviour**.

So what's the logic behind making default-configured instances shared without actually having a name when nothing else is shared?
Besides, assuming your proposal is in place, how do you aim to cover the use case when user wants default executor and wants it shared? You will have to do it manually (producer, builder) for no good reason.

The solution I have here follows the injection/configuration we have in place for all other cases. Ultimately _not having the config annotation is the same as having the annotation and placing the default values in it, right_?

njr-11 replied:
This concern (which Andy also addresses in a subsequent comment further down) will be discussed using comments on the main issue.  I added a summary there with my understanding of the two proposed options.

manovotn commented:
@njr-11 updated the PR with you comments, some rewording I took literally, some I adjusted a bit (leaving a comment to mark those), give it a spin if you will.

Also signed the ECA and added the signoff to commits (hence the force push).

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,156 @@
+
+Oftentimes there is a need to configure the executor instance to limit the number of asynchronous tasks that it can start or define which contexts to propagate.
+The `@ManagedExecutorConfig` annotation allows to configure all aspects of an injected `ManagedExecutor` instance.
+Note that using this annotation only, new instance of `ManagedExecutor` is created for every injection point even if those injection points have equivalent configuration.
```
Minor grammar updates:
allows to configure --> allows you to configure
Note that using --> Note that when using
new instance of --> a new instance of

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,156 @@
+
+The CDI qualifier, `@NamedInstance`, enables you to share the same instance of `ManagedExecutor` across multiple injection points.
+The `@NamedInstance` qualifier has a single parameter which specifies the unique name that is assigned to the given executor configuration.
+Having one injection point annotated with both `ManagedExecutorConfig` and `@NamedInstance`, all other injection points within the application that have a matching `@NamedInstance` qualifier will have the same `ManagedExecutor` instance injected.
```
Grammar:
Having one injection point annotated --> When one injection point is annotated

njr-11 commented on spec/src/main/asciidoc/cdi.asciidoc:
```diff
@@ -19,4 +19,156 @@
+
+Oftentimes there is a need to configure the `ThreadContext` instance to specify which contexts to propagate.
+The `@ThreadContextConfig` annotation allows to configure all aspects of an injected `ThreadContext` instance.
+Note that using this annotation only, new instance of `ThreadContext` is created for every injection point even if those injection points have equivalent configuration.
```
Minor grammar updates:
allows to configure --> allows you to configure
Note that using --> Note that when using
new instance of --> a new instance of

manovotn commented:
@njr-11 Either my English decayed drastically over last few months, or vast majority of your comments are not grammatical mistakes but rather stylistic differences.

manovotn commented:
@njr-11, @aguibert I have updated the PR as we agreed on last meeting, both the `overview` and the `cdi` part now contain similar two sentences about container providing instances per unique name unless a producer was specified.

With this I hope we are good to proceed.

manovotn commented:
Ok, it is a minor change for affecting all IPs would look like a bug in the first place.
Updated the PR accordingly to your suggestion and squashed into one commit.
